### PR TITLE
fix(core, windows): building improvements

### LINKED ISSUE #10931: 🐛 [firebase_core] Firebase Core 2.11.0 fails to build on Windows
## Bug report

**Describe the bug**
After adding `firebase_core` 2.11.0 to a Flutter project that supports Windows, the Windows build will fail to compile if built in release mode:

```
firebase_app.lib(d94e1bff3c0e8414a6679ea91c3e103e_firebase_app.dir_Debug_filesystem_desktop_windows.obj) : error LNK2001: unresolved external symbol __imp__invalid_parameter [C:\Git\firebase_test\build\windows\plugins\firebase_core\firebase_core_plugin.vcxproj]
firebase_app.lib(d94e1bff3c0e8414a6679ea91c3e103e_firebase_app.dir_Debug_variant.obj) : error LNK2001: unresolved external symbol __imp__invalid_parameter [C:\Git\firebase_test\build\windows\plugins\firebase_core\firebase_core_plugin.vcxproj]
firebase_app.lib(d94e1bff3c0e8414a6679ea91c3e103e_firebase_app.dir_Debug_variant_util.obj) : error LNK2001: unresolved external symbol __imp__invalid_parameter [C:\Git\firebase_test\build\windows\plugins\firebase_core\firebase_core_plugin.vcxproj]
firebase_app.lib(d3d7d08a438878e74aeb2cbaaedfc967_flatbuffers.dir_Debug_util.obj) : error LNK2001: unresolved external symbol __imp__invalid_parameter [C:\Git\firebase_test\build\windows\plugins\firebase_core\firebase_core_plugin.vcxproj]
firebase_app.lib(d94e1bff3c0e8414a6679ea91c3e103e_firebase_app.dir_Debug_reference_counted_future_impl.obj) : error LNK2001: unresolved external symbol __imp__invalid_parameter [C:\Git\firebase_test\build\windows\plugins\firebase_core\firebase_core_plugin.vcxproj]
firebase_app.lib(d94e1bff3c0e8414a6679ea91c3e103e_firebase_app.dir_Debug_util.obj) : error LNK2001: unresolved external symbol __imp__invalid_parameter [C:\Git\firebase_test\build\windows\plugins\firebase_core\firebase_core_plugin.vcxproj]
firebase_app.lib(d94e1bff3c0e8414a6679ea91c3e103e_firebase_app.dir_Debug_heartbeat_storage_desktop.obj) : error LNK2001: unresolved external symbol __imp__invalid_parameter [C:\Git\firebase_test\build\windows\plugins\firebase_core\firebase_core_plugin.vcxproj]
firebase_app.lib(d94e1bff3c0e8414a6679ea91c3e103e_firebase_app.dir_Debug_scheduler.obj) : error LNK2001: unresolved external symbol __imp__invalid_parameter [C:\Git\firebase_test\build\windows\plugins\firebase_core\firebase_core_plugin.vcxproj]
firebase_app.lib(d94e1bff3c0e8414a6679ea91c3e103e_firebase_app.dir_Debug_base64.obj) : error LNK2001: unresolved external symbol __imp__invalid_parameter [C:\Git\firebase_test\build\windows\plugins\firebase_core\firebase_core_plugin.vcxproj]
firebase_app.lib(d94e1bff3c0e8414a6679ea91c3e103e_firebase_app.dir_Debug_heartbeat_controller_desktop.obj) : error LNK2001: unresolved external symbol __imp__invalid_parameter [C:\Git\firebase_test\build\windows\plugins\firebase_core\firebase_core_plugin.vcxproj]
firebase_app.lib(d3d7d08a438878e74aeb2cbaaedfc967_flatbuffers.dir_Debug_idl_parser.obj) : error LNK2001: unresolved external symbol __imp__invalid_parameter [C:\Git\firebase_test\build\windows\plugins\firebase_core\firebase_core_plugin.vcxproj]
firebase_app.lib(d94e1bff3c0e8414a6679ea91c3e103e_firebase_app.dir_Debug_callback.obj) : error LNK2001: unresolved external symbol __imp__invalid_parameter [C:\Git\firebase_test\build\windows\plugins\firebase_core\firebase_core_plugin.vcxproj]
firebase_app.lib(d94e1bff3c0e8414a6679ea91c3e103e_firebase_app.dir_Debug_cleanup_notifier.obj) : error LNK2001: unresolved external symbol __imp__invalid_parameter [C:\Git\firebase_test\build\windows\plugins\firebase_core\firebase_core_plugin.vcxproj]
firebase_app.lib(d94e1bff3c0e8414a6679ea91c3e103e_firebase_app.dir_Debug_app_desktop.obj) : error LNK2001: unresolved external symbol __imp__invalid_parameter [C:\Git\firebase_test\build\windows\plugins\firebase_core\firebase_core_plugin.vcxproj]
firebase_app.lib(d94e1bff3c0e8414a6679ea91c3e103e_firebase_app.dir_Debug_app_options.obj) : error LNK2001: unresolved external symbol __imp__invalid_parameter [C:\Git\firebase_test\build\windows\plugins\firebase_core\firebase_core_plugin.vcxproj]
firebase_app.lib(d94e1bff3c0e8414a6679ea91c3e103e_firebase_app.dir_Debug_app_common.obj) : error LNK2001: unresolved external symbol __imp__invalid_parameter [C:\Git\firebase_test\build\windows\plugins\firebase_core\firebase_core_plugin.vcxproj]
firebase_app.lib(d94e1bff3c0e8414a6679ea91c3e103e_firebase_app.dir_Debug_date_provider.obj) : error LNK2001: unresolved external symbol __imp__invalid_parameter [C:\Git\firebase_test\build\windows\plugins\firebase_core\firebase_core_plugin.vcxproj]
firebase_app.lib(d94e1bff3c0e8414a6679ea91c3e103e_firebase_app.dir_Debug_filesystem_desktop_windows.obj) : error LNK2001: unresolved external symbol __imp__CrtDbgReport [C:\Git\firebase_test\build\windows\plugins\firebase_core\firebase_core_plugin.vcxproj]
firebase_app.lib(d94e1bff3c0e8414a6679ea91c3e103e_firebase_app.dir_Debug_variant.obj) : error LNK2001: unresolved external symbol __imp__CrtDbgReport [C:\Git\firebase_test\build\windows\plugins\firebase_core\firebase_core_plugin.vcxproj]
firebase_app.lib(d94e1bff3c0e8414a6679ea91c3e103e_firebase_app.dir_Debug_variant_util.obj) : error LNK2001: unresolved external symbol __imp__CrtDbgReport [C:\Git\firebase_test\build\windows\plugins\firebase_core\firebase_core_plugin.vcxproj]
firebase_app.lib(d3d7d08a438878e74aeb2cbaaedfc967_flatbuffers.dir_Debug_util.obj) : error LNK2001: unresolved external symbol __imp__CrtDbgReport [C:\Git\firebase_test\build\windows\plugins\firebase_core\firebase_core_plugin.vcxproj]
firebase_app.lib(d94e1bff3c0e8414a6679ea91c3e103e_firebase_app.dir_Debug_reference_counted_future_impl.obj) : error LNK2001: unresolved external symbol __imp__CrtDbgReport [C:\Git\firebase_test\build\windows\plugins\firebase_core\firebase_core_plugin.vcxproj]
firebase_app.lib(d94e1bff3c0e8414a6679ea91c3e103e_firebase_app.dir_Debug_util.obj) : error LNK2001: unresolved external symbol __imp__CrtDbgReport [C:\Git\firebase_test\build\windows\plugins\firebase_core\firebase_core_plugin.vcxproj]
firebase_app.lib(d94e1bff3c0e8414a6679ea91c3e103e_firebase_app.dir_Debug_heartbeat_storage_desktop.obj) : error LNK2001: unresolved external symbol __imp__CrtDbgReport [C:\Git\firebase_test\build\windows\plugins\firebase_core\firebase_core_plugin.vcxproj]
firebase_app.lib(d94e1bff3c0e8414a6679ea91c3e103e_firebase_app.dir_Debug_scheduler.obj) : error LNK2001: unresolved external symbol __imp__CrtDbgReport [C:\Git\firebase_test\build\windows\plugins\firebase_core\firebase_core_plugin.vcxproj]
firebase_app.lib(d94e1bff3c0e8414a6679ea91c3e103e_firebase_app.dir_Debug_base64.obj) : error LNK2001: unresolved external symbol __imp__CrtDbgReport [C:\Git\firebase_test\build\windows\plugins\firebase_core\firebase_core_plugin.vcxproj]
firebase_app.lib(d94e1bff3c0e8414a6679ea91c3e103e_firebase_app.dir_Debug_heartbeat_controller_desktop.obj) : error LNK2001: unresolved external symbol __imp__CrtDbgReport [C:\Git\firebase_test\build\windows\plugins\firebase_core\firebase_core_plugin.vcxproj]
firebase_app.lib(d3d7d08a438878e74aeb2cbaaedfc967_flatbuffers.dir_Debug_idl_parser.obj) : error LNK2001: unresolved external symbol __imp__CrtDbgReport [C:\Git\firebase_test\build\windows\plugins\firebase_core\firebase_core_plugin.vcxproj]
firebase_app.lib(d94e1bff3c0e8414a6679ea91c3e103e_firebase_app.dir_Debug_callback.obj) : error LNK2001: unresolved external symbol __imp__CrtDbgReport [C:\Git\firebase_test\build\windows\plugins\firebase_core\firebase_core_plugin.vcxproj]
firebase_app.lib(d94e1bff3c0e8414a6679ea91c3e103e_firebase_app.dir_Debug_cleanup_notifier.obj) : error LNK2001: unresolved external symbol __imp__CrtDbgReport [C:\Git\firebase_test\build\windows\plugins\firebase_core\firebase_core_plugin.vcxproj]
firebase_app.lib(d94e1bff3c0e8414a6679ea91c3e103e_firebase_app.dir_Debug_app_desktop.obj) : error LNK2001: unresolved external symbol __imp__CrtDbgReport [C:\Git\firebase_test\build\windows\plugins\firebase_core\firebase_core_plugin.vcxproj]
firebase_app.lib(d94e1bff3c0e8414a6679ea91c3e103e_firebase_app.dir_Debug_app_options.obj) : error LNK2001: unresolved external symbol __imp__CrtDbgReport [C:\Git\firebase_test\build\windows\plugins\firebase_core\firebase_core_plugin.vcxproj]
firebase_app.lib(d94e1bff3c0e8414a6679ea91c3e103e_firebase_app.dir_Debug_app_common.obj) : error LNK2001: unresolved external symbol __imp__CrtDbgReport [C:\Git\firebase_test\build\windows\plugins\firebase_core\firebase_core_plugin.vcxproj]firebase_app.lib(d94e1bff3c0e8414a6679ea91c3e103e_firebase_app.dir_Debug_date_provider.obj) : error LNK2001: unresolved external symbol __imp__CrtDbgReport [C:\Git\firebase_test\build\windows\plugins\firebase_core\firebase_core_plugin.vcxproj]
firebase_app.lib(d94e1bff3c0e8414a6679ea91c3e103e_firebase_app.dir_Debug_heartbeat_storage_desktop.obj) : error LNK2001: unresolved external symbol __imp__calloc_dbg [C:\Git\firebase_test\build\windows\plugins\firebase_core\firebase_core_plugin.vcxproj]
firebase_app.lib(d94e1bff3c0e8414a6679ea91c3e103e_firebase_app.dir_Debug_variant.obj) : error LNK2001: unresolved external symbol __imp__calloc_dbg [C:\Git\firebase_test\build\windows\plugins\firebase_core\firebase_core_plugin.vcxproj]
firebase_app.lib(d94e1bff3c0e8414a6679ea91c3e103e_firebase_app.dir_Debug_variant_util.obj) : error LNK2001: unresolved external symbol __imp__calloc_dbg [C:\Git\firebase_test\build\windows\plugins\firebase_core\firebase_core_plugin.vcxproj]firebase_app.lib(d3d7d08a438878e74aeb2cbaaedfc967_flatbuffers.dir_Debug_util.obj) : error LNK2001: unresolved external symbol __imp__calloc_dbg [C:\Git\firebase_test\build\windows\plugins\firebase_core\firebase_core_plugin.vcxproj]
firebase_app.lib(d94e1bff3c0e8414a6679ea91c3e103e_firebase_app.dir_Debug_app_options.obj) : error LNK2001: unresolved external symbol __imp__calloc_dbg [C:\Git\firebase_test\build\windows\plugins\firebase_core\firebase_core_plugin.vcxproj]
firebase_app.lib(d94e1bff3c0e8414a6679ea91c3e103e_firebase_app.dir_Debug_date_provider.obj) : error LNK2001: unresolved external symbol __imp__calloc_dbg [C:\Git\firebase_test\build\windows\plugins\firebase_core\firebase_core_plugin.vcxproj]
firebase_app.lib(d94e1bff3c0e8414a6679ea91c3e103e_firebase_app.dir_Debug_heartbeat_controller_desktop.obj) : error LNK2001: unresolved external symbol __imp__calloc_dbg [C:\Git\firebase_test\build\windows\plugins\firebase_core\firebase_core_plugin.vcxproj]
firebase_app.lib(d3d7d08a438878e74aeb2cbaaedfc967_flatbuffers.dir_Debug_idl_parser.obj) : error LNK2001: unresolved external symbol __imp__calloc_dbg [C:\Git\firebase_test\build\windows\plugins\firebase_core\firebase_core_plugin.vcxproj]
firebase_app.lib(d94e1bff3c0e8414a6679ea91c3e103e_firebase_app.dir_Debug_heartbeat_storage_desktop.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) public: static void * __cdecl std::_Crt_new_delete::operator new(unsigned __int64)" (__imp_??2_Crt_new_delete@std@@SAPEAX_K@Z) referenced in function "public: static unsigned __int64 __cdecl std::collate<char>::_Getcat(class std::locale::facet const * *,class std::locale const *)" (?_Getcat@?$collate@D@std@@SA_KPEAPEBVfacet@locale@2@PEBV42@@Z) [C:\Git\firebase_test\build\windows\plugins\firebase_core\firebase_core_plugin.vcxproj]
firebase_app.lib(d94e1bff3c0e8414a6679ea91c3e103e_firebase_app.dir_Debug_heartbeat_storage_desktop.obj) : error LNK2019: unresolved external symbol "__declspec(dllimport) public: static void __cdecl std::_Crt_new_delete::operator delete(void *)" (__imp_??3_Crt_new_delete@std@@SAXPEAX@Z) referenced in function "protected: virtual void * __cdecl std::collate<char>::`scalar deleting destructor'(unsigned int)" (??_G?$collate@D@std@@MEAAPEAXI@Z) [C:\Git\firebase_test\build\windows\plugins\firebase_core\firebase_core_plugin.vcxproj]
firebase_app.lib(d3d7d08a438878e74aeb2cbaaedfc967_flatbuffers.dir_Debug_util.obj) : error LNK2019: unresolved external symbol __imp__CrtSetReportFile referenced in function "void __cdecl f_b_flatbuffers::SetupDefaultCRTReportMode(void)" (?SetupDefaultCRTReportMode@f_b_flatbuffers@@YAXXZ) [C:\Git\firebase_test\build\windows\plugins\firebase_core\firebase_core_plugin.vcxproj]
firebase_app.lib(d3d7d08a438878e74aeb2cbaaedfc967_flatbuffers.dir_Debug_util.obj) : error LNK2019: unresolved external symbol __imp__CrtSetReportMode referenced in function "void __cdecl f_b_flatbuffers::SetupDefaultCRTReportMode(void)" (?SetupDefaultCRTReportMode@f_b_flatbuffers@@YAXXZ) [C:\Git\firebase_test\build\windows\plugins\firebase_core\firebase_core_plugin.vcxproj]
C:\Git\firebase_test\build\windows\plugins\firebase_core\Release\firebase_core_plugin.dll : fatal error LNK1120: 7 unresolved externals [C:\Git\firebase_test\build\windows\plugins\firebase_core\firebase_core_plugin.vcxproj]
```

### Steps to reproduce

1. Create a new Flutter project - `flutter create --platform windows firebase_test`
2. Add the dependency to pubspec.yaml:
```
dependencies:
  flutter:
    sdk: flutter
  firebase_core: 2.11.0
```
3. `flutter pub get`
4. `flutter run -d windows --release`
5. Observe the compile error

### Expected behavior

Firebase should not break builds on the Windows platform

### Sample project

https://github.com/insertjokehere/firebase-compile-error-windows

---

## Additional context

I've tried running `flutterfire configure`, but it seems like this tool does not support Windows builds yet?

---

### Flutter doctor

Run `flutter doctor` and paste the output below:

<details><summary>Click To Expand</summary>

```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 3.7.9, on Microsoft Windows [Version 10.0.22621.1555], locale en-NZ)
[✓] Windows Version (Installed version of Windows is version 10 or higher)
[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0-rc2)
[✓] Chrome - develop for the web
[✓] Visual Studio - develop for Windows (Visual Studio Professional 2022 17.2.4)
[✓] Android Studio (version 2021.1)
[✓] VS Code (version 1.78.0)
[✓] Connected device (4 available)
[✓] HTTP Host Availability

• No issues found!
```

</details>

---

### Flutter dependencies

Run `flutter pub deps -- --style=compact` and paste the output below:

<details><summary>Click To Expand</summary>

```
Dart SDK 2.19.6
Flutter SDK 3.7.9
firebase_test 1.0.0+1

dependencies:
- cupertino_icons 1.0.5
- firebase_core 2.11.0 [firebase_core_platform_interface firebase_core_web flutter meta]
- flutter 0.0.0 [characters collection js material_color_utilities meta vector_math sky_engine]

dev dependencies:
- flutter_lints 2.0.1 [lints]
- flutter_test 0.0.0 [flutter test_api path fake_async clock stack_trace vector_math async boolean_selector characters collection js matcher material_color_utilities meta source_span stream_channel string_scanner term_glyph]

transitive dependencies:
- async 2.10.0 [collection meta]
- boolean_selector 2.1.1 [source_span string_scanner]
- characters 1.2.1
- clock 1.1.1
- collection 1.17.0
- fake_async 1.3.1 [clock collection]
- firebase_core_platform_interface 4.7.0 [collection flutter flutter_test meta plugin_platform_interface]
- firebase_core_web 2.4.0 [firebase_core_platform_interface flutter flutter_web_plugins js meta]
- flutter_web_plugins 0.0.0 [flutter js characters collection material_color_utilities meta vector_math]
- js 0.6.5 [meta]
- lints 2.0.1
- matcher 0.12.13 [meta stack_trace]
- material_color_utilities 0.2.0
- meta 1.8.0
- path 1.8.2
- plugin_platform_interface 2.1.4 [meta]
- sky_engine 0.0.99
- source_span 1.9.1 [collection path term_glyph]
- stack_trace 1.11.0 [path]
- stream_channel 2.1.1 [async]
- string_scanner 1.2.0 [source_span]
- term_glyph 1.2.1
- test_api 0.4.16 [async boolean_selector collection meta source_span stack_trace stream_channel string_scanner term_glyph matcher]
- vector_math 2.1.4
```

</details>

---

### PR DESCRIPTION
**Pull request by viachaslavic**
## Description

1.  Fixed paths to SDK libraries depending on the build config.
2.  Using existing SDK on system according [desktop-workflow] to speed up the build. 

## Related Issues

fix #10931 

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`).
This will ensure a smooth and quick review process. Updating the `pubspec.yaml` and changelogs is not required.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [ ] All existing and new tests are passing.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] The analyzer (`melos run analyze`) does not report any problems on my PR.
- [ ] I read and followed the [Flutter Style Guide].
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change.
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/firebase/flutterfire/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[CLA]: https://cla.developers.google.com/
[desktop-workflow]: https://firebase.google.com/docs/cpp/setup?platform=ios#desktop-workflow

### CODE REVIEW

**Comment by Lyokone** [comment]
@viachaslavic thanks a lot for this PR. I'll review this quickly 👍

**Review by Lyokone — APPROVED** [review]

File: packages/firebase_core/firebase_core/windows/CMakeLists.txt [hunk 1]
```diff
@@ -6,27 +6,35 @@ cmake_minimum_required(VERSION 3.14)
...
+    file(SHA256 $ENV{FIREBASE_CPP_SDK_DIR}/file_hashes.txt FIREBASE_SDK_FILE_HASH)
 endif()
 
-if(NOT EXISTS ${EXTRACTED_PATH})
-    file(MAKE_DIRECTORY ${EXTRACTED_PATH})
-    file(ARCHIVE_EXTRACT INPUT ${FIREBASE_SDK_FILENAME}
-         DESTINATION ${EXTRACTED_PATH})
-endif()
-set(FIREBASE_CPP_SDK_DIR "${EXTRACTED_PATH}/firebase_cpp_sdk")
+if(FIREBASE_SDK_FILE_HASH STREQUAL "3163806e86ca534d417c0306fcf1068acb35dd7545f3ffdaa109f989b2fd411b")
```
Comment (russellwheatley) [inline packages/firebase_core/firebase_core/windows/CMakeLists.txt:14]: Excuse my ignorance, but is this hash based on the Firebase SDK version? i.e does this need to change every time we change the firebase version?

Comment (viachaslavic) [reply packages/firebase_core/firebase_core/windows/CMakeLists.txt:14]: @russellwheatley, yes.  I agree, this is not the most convenient and reliable approach. Instead this I'll try to parse the version from the version header.

Comment (viachaslavic) [reply packages/firebase_core/firebase_core/windows/CMakeLists.txt]: @russellwheatley Done.  Now the version is checked [natively](https://cmake.org/cmake/help/latest/variable/CMAKE_VERSION.html).

**Review by russellwheatley — APPROVED** [review]

**Comment by Lyokone** [comment]
Thanks a lot for the fix @viachaslavic

**Comment by insertjokehere** [comment]
Thanks @viachaslavic!

**Pull request merged**